File: school_library/library/page/my_library/my_library.py
import frappe


@frappe.whitelist()
def get_my_library_data():
    """Find the Library Member for the current user and return their issues."""
    user = frappe.session.user
    if not user or user == "Guest":
        frappe.throw("Please log in to view your library.")

    member = _find_member_for_user(user)
    if not member:
        return {
            "member": None,
            "active_issues": [],
            "history": [],
            "message": "No Library Member account found for you. Please contact the librarian."
        }

    active = frappe.get_all(
        "Book Issue",
        filters={"member": member.name, "return_date": ["is", "not set"]},
        fields=["name", "book", "book_title", "issue_date", "due_date", "status",
                "is_overdue", "overdue_days"],
        order_by="due_date asc",
        ignore_permissions=True
    )

    history = frappe.get_all(
        "Book Issue",
        filters={"member": member.name, "return_date": ["is", "set"]},
        fields=["name", "book", "book_title", "issue_date", "due_date",
                "return_date", "status", "is_overdue", "overdue_days"],
        order_by="issue_date desc",
        limit_page_length=50,
        ignore_permissions=True
    )

    return {
        "member": {
            "name": member.name,
            "member_name": member.member_name,
            "member_type": member.member_type,
            "max_books_allowed": member.max_books_allowed,
            "current_issued_count": member.current_issued_count,
            "is_active": member.is_active
        },
        "active_issues": active,
        "history": history
    }


def _find_member_for_user(user):
    """Match the logged-in user to a Library Member via their Student or Teacher record."""

    # Try Student via email_address
    student_name = frappe.db.get_value("Student", {"email_address": user}, "name")
    if student_name:
        member_name = frappe.db.get_value(
            "Library Member",
            {"member_type": "Student", "student": student_name, "is_active": 1},
            "name"
        )
        if member_name:
            return frappe.get_doc("Library Member", member_name)

    # Try Teacher via user link
    teacher_meta = frappe.get_meta("Teacher")
    user_field = None
    for df in teacher_meta.fields:
        if df.fieldtype == "Link" and df.options == "User":
            user_field = df.fieldname
            break
    if user_field:
        teacher_name = frappe.db.get_value("Teacher", {user_field: user}, "name")
        if teacher_name:
            member_name = frappe.db.get_value(
                "Library Member",
                {"member_type": "Teacher", "teacher": teacher_name, "is_active": 1},
                "name"
            )
            if member_name:
                return frappe.get_doc("Library Member", member_name)

    # Fallback — External member by email
    member_name = frappe.db.get_value(
        "Library Member",
        {"member_type": "External", "external_email": user, "is_active": 1},
        "name"
    )
    if member_name:
        return frappe.get_doc("Library Member", member_name)

    return None

File: school_library/library/page/my_library/my_library.js
frappe.pages['my-library'].on_page_load = function(wrapper) {
    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'My Library',
        single_column: true
    });

    const $body = $(`
        <div class="my-lib-body">
            <style>
                .ml-card { background:#fff; border:1px solid #e2e8f0; border-radius:8px; padding:18px; margin-bottom:16px; }
                .ml-header { display:flex; justify-content:space-between; align-items:center; flex-wrap:wrap; gap:10px; }
                .ml-name { font-size:18px; font-weight:600; color:#1e293b; }
                .ml-meta { font-size:13px; color:#64748b; }
                .ml-stats { display:flex; gap:14px; flex-wrap:wrap; }
                .ml-stat { background:#f8fafc; padding:8px 14px; border-radius:6px; font-size:13px; }
                .ml-stat b { color:#1e293b; font-size:16px; }
                .ml-section h4 { margin:0 0 12px; color:#1e293b; }
                .ml-table { width:100%; border-collapse:collapse; }
                .ml-table th { background:#f8fafc; padding:8px 10px; text-align:left; font-size:11px; color:#64748b; text-transform:uppercase; letter-spacing:0.4px; border-bottom:1px solid #e2e8f0; }
                .ml-table td { padding:10px; font-size:13px; border-bottom:1px solid #f1f5f9; }
                .ml-empty { padding:18px; text-align:center; color:#94a3b8; font-style:italic; }
                .badge { padding:2px 8px; border-radius:10px; font-size:11px; font-weight:500; }
                .badge-issued { background:#dbeafe; color:#1e40af; }
                .badge-returned { background:#d1fae5; color:#065f46; }
                .badge-overdue { background:#fee2e2; color:#991b1b; }
                .badge-lost { background:#f3f4f6; color:#374151; }
                .ml-banner-warning { background:#fef3c7; color:#92400e; padding:12px; border-radius:6px; margin-bottom:14px; }
            </style>

            <div id="ml-content">
                <div class="ml-card"><div class="ml-empty">Loading...</div></div>
            </div>
        </div>
    `).appendTo(page.body);

    load_my_library();
};

function load_my_library() {
    frappe.call({
        method: 'school_library.library.page.my_library.my_library.get_my_library_data',
        callback: function(r) {
            const $c = $('#ml-content').empty();
            if (!r.message) return;
            const data = r.message;

            if (!data.member) {
                $c.html(`<div class="ml-card"><div class="ml-banner-warning">
                    ${frappe.utils.escape_html(data.message || 'No library account found.')}
                </div></div>`);
                return;
            }

            // Member info
            const m = data.member;
            $c.append(`
                <div class="ml-card">
                    <div class="ml-header">
                        <div>
                            <div class="ml-name">${frappe.utils.escape_html(m.member_name)}</div>
                            <div class="ml-meta">${m.member_type} · ${m.name}</div>
                        </div>
                        <div class="ml-stats">
                            <div class="ml-stat">Issued <b>${m.current_issued_count || 0}</b></div>
                            <div class="ml-stat">Limit <b>${m.max_books_allowed || 0}</b></div>
                        </div>
                    </div>
                </div>
            `);

            // Active issues
            let activeHtml = '<div class="ml-card ml-section"><h4>Currently Issued</h4>';
            if (!data.active_issues || !data.active_issues.length) {
                activeHtml += '<div class="ml-empty">You have no books issued right now.</div>';
            } else {
                activeHtml += `<table class="ml-table"><thead><tr>
                    <th>Book</th><th>Issued</th><th>Due</th><th>Status</th>
                </tr></thead><tbody>`;
                data.active_issues.forEach(row => {
                    const statusClass = row.is_overdue ? 'badge-overdue' : 'badge-issued';
                    const statusLabel = row.is_overdue ? `Overdue (${row.overdue_days}d)` : 'Issued';
                    activeHtml += `<tr>
                        <td>${frappe.utils.escape_html(row.book_title || '')}</td>
                        <td>${frappe.datetime.str_to_user(row.issue_date)}</td>
                        <td>${frappe.datetime.str_to_user(row.due_date)}</td>
                        <td><span class="badge ${statusClass}">${statusLabel}</span></td>
                    </tr>`;
                });
                activeHtml += '</tbody></table>';
            }
            activeHtml += '</div>';
            $c.append(activeHtml);

            // History
            let histHtml = '<div class="ml-card ml-section"><h4>History</h4>';
            if (!data.history || !data.history.length) {
                histHtml += '<div class="ml-empty">No past book issues yet.</div>';
            } else {
                histHtml += `<table class="ml-table"><thead><tr>
                    <th>Book</th><th>Issued</th><th>Due</th><th>Returned</th><th>Status</th>
                </tr></thead><tbody>`;
                data.history.forEach(row => {
                    let badge = 'badge-returned';
                    let label = 'Returned';
                    if (row.status === 'Lost') { badge = 'badge-lost'; label = 'Lost'; }
                    if (row.is_overdue && row.status === 'Returned') {
                        label = `Returned late (${row.overdue_days}d)`;
                    }
                    histHtml += `<tr>
                        <td>${frappe.utils.escape_html(row.book_title || '')}</td>
                        <td>${frappe.datetime.str_to_user(row.issue_date)}</td>
                        <td>${frappe.datetime.str_to_user(row.due_date)}</td>
                        <td>${row.return_date ? frappe.datetime.str_to_user(row.return_date) : '—'}</td>
                        <td><span class="badge ${badge}">${label}</span></td>
                    </tr>`;
                });
                histHtml += '</tbody></table>';
            }
            histHtml += '</div>';
            $c.append(histHtml);
        }
    });
}
